cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java

```diff
@@ -156,9 +156,26 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
     return dataset;
   }
 
+  @Nullable
+  @Override
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetInstanceId, @Nullable Map<String, String> arguments,
+                                          @Nullable ClassLoader classLoader,
+                                          DatasetClassLoaderProvider classLoaderProvider,
+                                          @Nullable Iterable<? extends Id> owners, AccessType accessType)
+    throws DatasetManagementException, IOException {
+    T dataset = super.getDataset(datasetInstanceId, arguments, classLoader, classLoaderProvider, owners, accessType);
+    writeLineage(datasetInstanceId, dataset, accessType);
```

Comment:
hum.. nevermind.. I see the check is actually inside the `writeLineage` method, although the `writeLineage` method should have the `T dataset` parameter annotated with `@Nullable`.